lib/ansible/plugins/httpapi/ftd.py

```diff
@@ -2,175 +2,213 @@
+
             headers = self._authorized_headers()
             headers['Content-Type'] = content_type
             headers['Content-Length'] = len(body)
-            response, response_data = self.connection.send(
-                url, data=body, method='POST', headers=headers
-            )
-            try:
-                ret = json.loads(to_text(response_data.getvalue()))
-            except:
-                raise ConnectionError('Response was not valid JSON, got {0}'
-                                      .format(response_data.getvalue()))
-            return ret
-
-    def download_file(self, from_url, to_path):
-        url = construct_url_path(from_url)
+
+            _, response_data = self.connection.send(url, data=body, method=HTTPMethod.POST, headers=headers)
+            return self._response_to_json(response_data.getvalue())
+
+    def download_file(self, from_url, to_path, path_params=None):
+        url = construct_url_path(from_url, path_params=path_params)
         response, response_data = self.connection.send(
-            url, data=None, method='GET',
+            url, data=None, method=HTTPMethod.GET,
             headers=self._authorized_headers()
         )
+
         if os.path.isdir(to_path):
             filename = extract_filename_from_headers(response.info())
             to_path = os.path.join(to_path, filename)
 
         with open(to_path, "wb") as output_file:
-            output_file.write(to_text(response_data.getvalue()))
-
-    def update_auth(self, response, response_data):
-        return None
-
-    def _set_token_info(self, response_data):
-        try:
-            token_info = json.loads(to_text(response_data.getvalue()))
-        except ValueError:
-            raise
-        if 'refresh_token' in token_info:
-            self.refresh_token = token_info['refresh_token']
-        if 'access_token' in token_info:
-            self.access_token = token_info['access_token']
+            output_file.write(response_data.getvalue())
 
     def handle_httperror(self, exc):
-        # Called by connection plugin when it gets HTTP Error for a request.
-        # Connection plugin will resend this request if we return true here.
-        if (exc.code == TOKEN_EXPIRATION_STATUS_CODE or
-           exc.code == UNAUTHORIZED_STATUS_CODE):
-            # Stored auth appears to be invalid, clear and retry
+        # None means that the exception will be passed further to the caller
+        retry_request = None
+        if exc.code == TOKEN_EXPIRATION_STATUS_CODE or exc.code == UNAUTHORIZED_STATUS_CODE:
             self.connection._auth = None
-            self.login(self.connection.get_option('remote_user'),
-                       self.connection.get_option('password'))
-            return True
-
-        return False
+            self.login(self.connection.get_option('remote_user'), self.connection.get_option('password'))
+            retry_request = True
+        return retry_request
```

Comment:
This can be converted to return True. No need of new variable retry_request

Reply:
Thanks, done.